### ISSUE TITLE
Wrong Content-Disposition handling with japanese chars

### ISSUE THREAD

**Issue by The-Compiler** [The-Compiler/qutebrowser#1724]
When downloading http://sukebei.nyaa.se/?page=view&tid=348759 we get " [131129][é­äºº]é»ç£ ï½æ°é«ãèå¥³ã¯ç½æ¿ã«æã¾ãï½ ã«ã°ã¤Ãã«ã¼ã»ã«ã¼Ããã¤ã¢ æèã«è ¢ãè²ãçãçãéç£éã®å¤ç·¨.mp4.torrent" as proposed filename, while Firefox/Chromium seem to handle it fine.

**Comment by Kingdread** [comment]
It seems like the sent Content-Disposition header is `b'inline; filename="[131129][\xe9\xad\x94\xe4\xba\xba]\xe9\xbb\x92\xe7\x8d\xa3 \xef\xbd\x9e\xe6\xb0\x97\xe9\xab\x98\xe3\x81\x8d\xe8\x81\x96\xe5\xa5\xb3\xe3\x81\xaf\xe7\x99\xbd\xe6\xbf\x81\xe3\x81\xab\xe6\x9f\x93\xe3\x81\xbe\xe3\x82\x8b\xef\xbd\x9e \xe3\x82\xab\xe3\x82\xb0\xe3\x83\xa4\xc3\x97\xe3\x83\xab\xe3\x83\xbc\xe3\x83\xbb\xe3\x83\xab\xe3\x83\xbc\xc3\x97\xe3\x83\x9e\xe3\x82\xa4\xe3\x82\xa2 \xe6\x9f\x94\xe8\x82\x8c\xe3\x81\xab\xe8\xa0\xa2\xe3\x81\x8f\xe8\x9f\xb2\xe3\x80\x81\xe7\x8c\x9b\xe3\x82\x8a\xe7\x8b\x82\xe3\x81\x86\xe9\x87\x8e\xe7\x8d\xa3\xe9\x81\x94\xe3\x81\xae\xe5\xa4\x9c\xe7\xb7\xa8.mp4.torrent"`, which is the (correct) filename encoded as UTF-8. This is however non-standard, as the header should be encoded in ISO-8859-1, and it's the same case as [this test case](https://greenbytes.de/tech/tc2231/#attwithutf8fnplain) from a RFC 6266 test suite. qutebrowser always decodes as ISO-8859-1 (`qutebrowser/browser/webkit/rfc6266.py`  line 298), in accordance with [RFC 6266](https://tools.ietf.org/html/rfc6266). The RFC also meantions ["encoding sniffing"](https://tools.ietf.org/html/rfc6266#appendix-C.3) and says

> As with the approaches above, this is not interoperable and,
> furthermore, risks misinterpreting the actual value.

Firefox/Chromium do some kind of encoding sniffing, which is why they produce the correct filename (and fail the test in the test suite).

The fix should be easy, the question is just do we
* stick to the standard, even when some webservers/scripts don't
* do what Firefox/Chromium do and try to interpret the value as UTF-8, even if it is nonstandard and may produce incorrect results

**Comment by The-Compiler** [comment]
Thanks for the investigation!

So in theory I'd rather have interoperability with "broken" websites and other browsers than 100% conforming to the spec. I find "risks misinterpreting the actual value" a bit ironic here, because that's what happened :wink:

However, I'm not sure if there is a straightforward solution for this short of using a new dependency ([chardet](https://pypi.python.org/pypi/chardet)). That probably isn't worth it, and with QtWebEngine the content-disposition header will be handled by Chromium if I'm not mistaken.

If you agree, feel free to close this. If you find a nice short solution, let me know :laughing:

**Comment by Kingdread** [comment]
How would you feel if your `foo-Ã¤.html` filename was misinterpreted by every browser as `foo-ä.html`? The RFC authors probably wanted to prevent that :wink:

The only short (hackish) solution I know would be something like

```python
try:
    value = value.decode("utf-8")
except UnicodeDecodeError:
    value = value.decode("iso-8859-1")
```

but that only covers one special case and I don't know how often it really provides better/worse results.

If this will be covered by QtWebEngine in the future anyway, it's probably really not worth.

tl;dr: I agree :laughing:
